File: kangtaku/nonDecreasingArray.java
class Solution {
    public boolean checkPossibility(int[] nums) {
        if (nums.length == 1) return true;
        int n = nums.length;
        for (int i = 0; i < n - 1; i++) {
            if (nums[i] > nums[i + 1]) {
                int temp1 = nums[i + 1];
                int temp2 = nums[i];
                
                nums[i + 1] = temp2;
                boolean isOkay = true;
                for (int j = 0; j < n - 1; j++) {
                    if (nums[j] > nums[j + 1]) {
                        isOkay = false;
                        break;
                    }
                }
                if (isOkay) {
                    return true;
                }
                isOkay = true;
                nums[i] = temp1;
                nums[i + 1] = temp1;
                for (int j = 0; j < n - 1; j++) {
                    if (nums[j] > nums[j + 1]) {
                        isOkay = false;
                        break;
                    }
                }
                if (isOkay) {
                    return true;
                }
                return false;
            }
        }
        return true;
    }
}

File: byssup/array/Non-decreasing Array.py
class Solution(object):
    def checkPossibility(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        i, j = self.get_idx(nums)
        if not i and not j:
            return True
        for k in [i, j]:
            ii, jj = self.get_idx(nums[:k] + nums[k + 1:])
            if not ii and not jj:
                return True
        return False

    def get_idx(self, nums):
        i = 0
        for j, num in enumerate(nums):
            if nums[i] > num:
                return i, j
            i = j
        return None, None
